src/gentropy/method/window_based_clumping.py

```diff
@@ -175,161 +180,128 @@ def clump(
+                        )
                     ),
-                )
-                # Within each cluster variants are ranked by significance:
-                .withColumn("pvRank", f.row_number().over(cluster_window))
-                # Collect positions in cluster for the most significant variant (complexity reduction):
-                .withColumn(
-                    "collectedPositions",
-                    f.when(
-                        f.col("pvRank") == 1,
-                        f.collect_list(f.col("position")).over(
-                            cluster_window.rowsBetween(
-                                Window.currentRow, Window.unboundedFollowing
-                            )
-                        ),
-                    ).otherwise(f.array()),
-                )
-                # Get semi indices only ONCE per cluster:
-                .withColumn(
-                    "semiIndices",
-                    f.when(
-                        f.size(f.col("collectedPositions")) > 0,
-                        fml.vector_to_array(
-                            f.udf(WindowBasedClumping._prune_peak, VectorUDT())(
-                                fml.array_to_vector(f.col("collectedPositions")),
-                                f.lit(window_length),
-                            )
-                        ),
+                ).otherwise(f.array()),
+            )
+            # Get semi indices only ONCE per cluster:
+            .withColumn(
+                "semiIndices",
+                f.when(
+                    f.size(f.col("collectedPositions")) > 0,
+                    fml.vector_to_array(
+                        f.udf(WindowBasedClumping._prune_peak, VectorUDT())(
+                            fml.array_to_vector(f.col("collectedPositions")),
+                            f.lit(distance),
+                        )
                     ),
+                ),
+            )
+            # Propagating the result of the above calculation for all rows:
+            .withColumn(
+                "semiIndices",
+                f.when(
+                    f.col("semiIndices").isNull(),
+                    f.first(f.col("semiIndices"), ignorenulls=True).over(
+                        cluster_window
+                    ),
+                ).otherwise(f.col("semiIndices")),
+            )
+            # Keeping semi indices only:
+            .filter(f.col("semiIndices")[f.col("pvRank") - 1] > 0)
+            .drop("pvRank", "collectedPositions", "semiIndices", "cluster_id")
+            # Adding study-locus id:
+            .withColumn(
+                "studyLocusId",
+                StudyLocus.assign_study_locus_id(f.col("studyId"), f.col("variantId")),
+            )
+            # Initialize QC column as array of strings:
+            .withColumn("qualityControls", f.array().cast(t.ArrayType(t.StringType())))
+        ).persist()
+
+        if collect_locus:
```

Comment:
To my tast the indentation is a bit too much, I would move this logic to a different funciton.

```python
if collect_locus:
    if not collect_locus_distance:
        collect_locus_distance = distance
    locus_df = cls._locus_collection( clumped_dataframe, summary_statistics, collect_distance)
```

But this is mostly stylistic.

Reply:
This function is a bit too long for my taste as well. I'll add the function

Reply:
When you have time have a quick look at the new structure. I have doubts about the `_annotate_locus_using_summary_statistics` (which refers to the same logic you were referring to as `_locus_collection`). Do you think this function belongs to `StudyLocus`? It doesn't look very clumping-specific anymore.

Reply:
That's a very good point. Because in a sense, if we have the semi indices (which are SL), the locus collection is applied on SL. It also provides some extra flexibilty in terms of allowing a more granular control over which SL you want to collect for finemapping. This would be a great addition.

```python
(
	summary_statistics
	.window_based_clumping() # Returns SL
	.collect_locus(summary_statistics) 
)
```